airtel/client.go

```diff
@@ -0,0 +1,195 @@
+	base, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if !strings.HasSuffix(base.Path, "/") {
+		base.Path = base.Path + "/"
+	}
+
+	retryTransport := rehttp.NewTransport(
+		http.DefaultTransport,
+		rehttp.RetryAll(
+			rehttp.RetryMaxRetries(retries),
+			rehttp.RetryAny(
+				rehttp.RetryTemporaryErr(),
+				rehttp.RetryStatuses(502, 503),
+			),
+		),
+		rehttp.ExpJitterDelay(100*time.Millisecond, 1*time.Second),
+	)
+
+	httpClient := &http.Client{
+		Transport: retryTransport,
+	}
+
+	inner := &client.Client{
+		Client:    httpClient,
+		BaseURL:   base,
+		UserAgent: userAgent,
+	}
+
+	client := new(Client)
+	client.inner = inner
+
+	client.BaseURL = base
+	client.Country = cfg.Country
+	client.Currency = cfg.Currency
+	client.EncryptedPin = cfg.Pin
+	client.ClientId = cfg.ClientId
+	client.ClientSceret = cfg.Secret
+	client.GrantType = cfg.Grant
+	client.TokenSource = source
+
+	if client.TokenSource == nil {
+		client.TokenSource = newTokenSource(client, cfg)
+	}
+	return client, nil
+}
+
+// NewDefault returns a new AIRTEL API client using the http.Client
+// But it take payment credential parameter
+// default "https://openapi.airtel.africa" address, country RW(Rwanda) and RWF(Rwandan franc).
+func NewDefault(pin, clientId, secret, grant string) (*Client, error) {
+	config := &Config{
+		Pin:      pin,
+		ClientId: clientId,
+		Secret:   secret,
+		Grant:    grant,
+		Currency: currency,
+		Country:  country,
+	}
+	return New(config, nil, baseUrl, false, defaultRetries)
+}
+
+// do wraps the Client.Do function by creating the Request and
+// unmarshalling the response according to user expected output.
+func (c *Client) do(ctx context.Context, method, path string, in, out interface{}, headers http.Header) (*client.Response, error) {
+	req := &client.Request{
+		Method: method,
+		Path:   path,
+		Header: make(http.Header),
+	}
+
+	// if we are posting or putting data, we need to
+	// write it to the body of the request.
+	if in != nil {
+		buf := new(bytes.Buffer)
+		_ = json.NewEncoder(buf).Encode(in)
+		req.Header = map[string][]string{
+			"Content-Type": {"application/json"},
+		}
+		req.Body = buf
+	}
+
+	// set the request headers
+
+	if headers != nil {
+		for k, v := range headers {
+			req.Header.Set(k, v[0])
+		}
+	}
+
+	if c.inner.UserAgent != "" {
+		req.Header.Set("User-Agent", c.inner.UserAgent)
+	}
+
+	// set auth token from TokenSource
+	token, err := c.TokenSource.Token(ctx)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Authorization", "Bearer "+token.Token)
+
+	// execute the http request using airtel.Client.Do()
+	res, err := c.inner.Do(ctx, req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+
+	// if an error is encountered, unmarshal and return the
+	// error response.
+	switch res.Status {
+	case http.StatusUnauthorized:
+		_, err = c.TokenSource.Token(ctx)
+		if err != nil {
+			return nil, err
+		}
+		c.do(ctx, method, path, in, out, headers)
```

Comment:
You are using the wrong `do` method, this going to recursively call itself until you are out of memory.

Reply:
no cause for the second attempt token will be set the our will also changed from `401`,
also how about I do it like this `return c.do(ctx, method, path, in, out, headers)` ?